### ISSUE TITLE
PD E2E tests fail in soak with `Pod cannot be started due to exceeded capacity`

### ISSUE THREAD

**Issue by saad-ali** [GoogleCloudPlatform/kubernetes#10899]
PD tests fail with `Pod cannot be started due to exceeded capacity` in continuous soak-test

e.g. jenkins: /job/kubernetes-soak-continuous-e2e/1845/

```
Pod Disks should schedule a pod w/ a RW PD, remove it, then schedule it on another host

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pd.go:113
Expected error:
    <*errors.errorString | 0xc2086d4620>: {
        s: "Giving up; pod went into failed status: \n(*api.Pod){TypeMeta:(api.TypeMeta){Kind:(string) APIVersion:(string)} ObjectMeta:(api.ObjectMeta){Name:(string)pd-test-ab871a3d-2483-11e5-b9d9-42010af01555 GenerateName:(string) Namespace:(string)default SelfLink:(string)/api/v1/namespaces/default/pods/pd-test-ab871a3d-2483-11e5-b9d9-42010af01555 UID:(types.UID)b186db25-2483-11e5-a4e3-42010af01857 ResourceVersion:(string)12732 Generation:(int64)0 CreationTimestamp:(util.Time)2015-07-07 01:39:34 -0700 PDT DeletionTimestamp:(*util.Time)<nil> Labels:(map[string]string)<nil> Annotations:(map[string]string)<nil>} Spec:(api.PodSpec){Volumes:([]api.Volume)[{Name:(string)testpd VolumeSource:(api.VolumeSource){HostPath:(*api.HostPathVolumeSource)<nil> EmptyDir:(*api.EmptyDirVolumeSource)<nil> GCEPersistentDisk:(*api.GCEPersistentDiskVolumeSource){PDName:(string)e2e-soak-a72419da-2483-11e5-b9d9-42010af01555 FSType:(string)ext4 Partition:(int)0 ReadOnly:(bool)false} AWSElasticBlockStore:(*api.AWSElasticBlockStoreVolumeSource)<nil> GitRepo:(*api.GitRepoVolumeSource)<nil> Secret:(*api.SecretVolumeSource)<nil> NFS:(*api.NFSVolumeSource)<nil> ISCSI:(*api.ISCSIVolumeSource)<nil> Glusterfs:(*api.GlusterfsVolumeSource)<nil> PersistentVolumeClaim:(*api.PersistentVolumeClaimVolumeSource)<nil> RBD:(*api.RBDVolumeSource)<nil>}} {Name:(string)default-token-812kd VolumeSource:(api.VolumeSource){HostPath:(*api.HostPathVolumeSource)<nil> EmptyDir:(*api.EmptyDirVolumeSource)<nil> GCEPersistentDisk:(*api.GCEPersistentDiskVolumeSource)<nil> AWSElasticBlockStore:(*api.AWSElasticBlockStoreVolumeSource)<nil> GitRepo:(*api.GitRepoVolumeSource)<nil> Secret:(*api.SecretVolumeSource){SecretName:(string)default-token-812kd} NFS:(*api.NFSVolumeSource)<nil> ISCSI:(*api.ISCSIVolumeSource)<nil> Glusterfs:(*api.GlusterfsVolumeSource)<nil> PersistentVolumeClaim:(*api.PersistentVolumeClaimVolumeSource)<nil> RBD:(*api.RBDVolumeSource)<nil>}}] Containers:([]api.Container)[{Name:(string)testpd Image:(string)gcr.io/google_containers/pause Command:([]string)<nil> Args:([]string)<nil> WorkingDir:(string) Ports:([]api.ContainerPort)<nil> Env:([]api.EnvVar)<nil> Resources:(api.ResourceRequirements){Limits:(api.ResourceList)map[cpu:100m] Requests:(api.ResourceList)<nil>} VolumeMounts:([]api.VolumeMount)[{Name:(string)testpd ReadOnly:(bool)false MountPath:(string)/testpd} {Name:(string)default-token-812kd ReadOnly:(bool)true MountPath:(string)/var/run/secrets/kubernetes.io/serviceaccount}] LivenessProbe:(*api.Probe)<nil> ReadinessProbe:(*api.Probe)<nil> Lifecycle:(*api.Lifecycle)<nil> TerminationMessagePath:(string)/dev/termination-log ImagePullPolicy:(api.PullPolicy)IfNotPresent SecurityContext:(*api.SecurityContext)<nil>}] RestartPolicy:(api.RestartPolicy)Always TerminationGracePeriodSeconds:(*int64)<nil> ActiveDeadlineSeconds:(*int64)<nil> DNSPolicy:(api.DNSPolicy)ClusterFirst NodeSelector:(map[string]string)<nil> ServiceAccountName:(string)default NodeName:(string)e2e-soak-minion-urqn HostNetwork:(bool)false ImagePullSecrets:([]api.LocalObjectReference)<nil>} Status:(api.PodStatus){Phase:(api.PodPhase)Failed Conditions:([]api.PodCondition)<nil> Message:(string)Pod cannot be started due to exceeded capacity Reason:(string)CapacityExceeded HostIP:(string) PodIP:(string) StartTime:(*util.Time)2015-07-07 01:39:34 -0700 PDT ContainerStatuses:([]api.ContainerStatus)<nil>}}",
    }
    Giving up; pod went into failed status: 
    (*api.Pod){TypeMeta:(api.TypeMeta){Kind:(string) APIVersion:(string)} ObjectMeta:(api.ObjectMeta){Name:(string)pd-test-ab871a3d-2483-11e5-b9d9-42010af01555 GenerateName:(string) Namespace:(string)default SelfLink:(string)/api/v1/namespaces/default/pods/pd-test-ab871a3d-2483-11e5-b9d9-42010af01555 UID:(types.UID)b186db25-2483-11e5-a4e3-42010af01857 ResourceVersion:(string)12732 Generation:(int64)0 CreationTimestamp:(util.Time)2015-07-07 01:39:34 -0700 PDT DeletionTimestamp:(*util.Time)<nil> Labels:(map[string]string)<nil> Annotations:(map[string]string)<nil>} Spec:(api.PodSpec){Volumes:([]api.Volume)[{Name:(string)testpd VolumeSource:(api.VolumeSource){HostPath:(*api.HostPathVolumeSource)<nil> EmptyDir:(*api.EmptyDirVolumeSource)<nil> GCEPersistentDisk:(*api.GCEPersistentDiskVolumeSource){PDName:(string)e2e-soak-a72419da-2483-11e5-b9d9-42010af01555 FSType:(string)ext4 Partition:(int)0 ReadOnly:(bool)false} AWSElasticBlockStore:(*api.AWSElasticBlockStoreVolumeSource)<nil> GitRepo:(*api.GitRepoVolumeSource)<nil> Secret:(*api.SecretVolumeSource)<nil> NFS:(*api.NFSVolumeSource)<nil> ISCSI:(*api.ISCSIVolumeSource)<nil> Glusterfs:(*api.GlusterfsVolumeSource)<nil> PersistentVolumeClaim:(*api.PersistentVolumeClaimVolumeSource)<nil> RBD:(*api.RBDVolumeSource)<nil>}} {Name:(string)default-token-812kd VolumeSource:(api.VolumeSource){HostPath:(*api.HostPathVolumeSource)<nil> EmptyDir:(*api.EmptyDirVolumeSource)<nil> GCEPersistentDisk:(*api.GCEPersistentDiskVolumeSource)<nil> AWSElasticBlockStore:(*api.AWSElasticBlockStoreVolumeSource)<nil> GitRepo:(*api.GitRepoVolumeSource)<nil> Secret:(*api.SecretVolumeSource){SecretName:(string)default-token-812kd} NFS:(*api.NFSVolumeSource)<nil> ISCSI:(*api.ISCSIVolumeSource)<nil> Glusterfs:(*api.GlusterfsVolumeSource)<nil> PersistentVolumeClaim:(*api.PersistentVolumeClaimVolumeSource)<nil> RBD:(*api.RBDVolumeSource)<nil>}}] Containers:([]api.Container)[{Name:(string)testpd Image:(string)gcr.io/google_containers/pause Command:([]string)<nil> Args:([]string)<nil> WorkingDir:(string) Ports:([]api.ContainerPort)<nil> Env:([]api.EnvVar)<nil> Resources:(api.ResourceRequirements){Limits:(api.ResourceList)map[cpu:100m] Requests:(api.ResourceList)<nil>} VolumeMounts:([]api.VolumeMount)[{Name:(string)testpd ReadOnly:(bool)false MountPath:(string)/testpd} {Name:(string)default-token-812kd ReadOnly:(bool)true MountPath:(string)/var/run/secrets/kubernetes.io/serviceaccount}] LivenessProbe:(*api.Probe)<nil> ReadinessProbe:(*api.Probe)<nil> Lifecycle:(*api.Lifecycle)<nil> TerminationMessagePath:(string)/dev/termination-log ImagePullPolicy:(api.PullPolicy)IfNotPresent SecurityContext:(*api.SecurityContext)<nil>}] RestartPolicy:(api.RestartPolicy)Always TerminationGracePeriodSeconds:(*int64)<nil> ActiveDeadlineSeconds:(*int64)<nil> DNSPolicy:(api.DNSPolicy)ClusterFirst NodeSelector:(map[string]string)<nil> ServiceAccountName:(string)default NodeName:(string)e2e-soak-minion-urqn HostNetwork:(bool)false ImagePullSecrets:([]api.LocalObjectReference)<nil>} Status:(api.PodStatus){Phase:(api.PodPhase)Failed Conditions:([]api.PodCondition)<nil> Message:(string)Pod cannot be started due to exceeded capacity Reason:(string)CapacityExceeded HostIP:(string) PodIP:(string) StartTime:(*util.Time)2015-07-07 01:39:34 -0700 PDT ContainerStatuses:([]api.ContainerStatus)<nil>}}
not to have occurred


Pod Disks should schedule a pod w/ a readonly PD on two hosts, then remove both.

/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pd.go:172
Expected error:
    <*errors.errorString | 0xc209024410>: {
        s: "Giving up; pod went into failed status: \n(*api.Pod){TypeMeta:(api.TypeMeta){Kind:(string) APIVersion:(string)} ObjectMeta:(api.ObjectMeta){Name:(string)pd-test-4e63f3d7-2487-11e5-b9d9-42010af01555 GenerateName:(string) Namespace:(string)default SelfLink:(string)/api/v1/namespaces/default/pods/pd-test-4e63f3d7-2487-11e5-b9d9-42010af01555 UID:(types.UID)60556283-2487-11e5-a4e3-42010af01857 ResourceVersion:(string)15075 Generation:(int64)0 CreationTimestamp:(util.Time)2015-07-07 02:05:56 -0700 PDT DeletionTimestamp:(*util.Time)<nil> Labels:(map[string]string)<nil> Annotations:(map[string]string)<nil>} Spec:(api.PodSpec){Volumes:([]api.Volume)[{Name:(string)testpd VolumeSource:(api.VolumeSource){HostPath:(*api.HostPathVolumeSource)<nil> EmptyDir:(*api.EmptyDirVolumeSource)<nil> GCEPersistentDisk:(*api.GCEPersistentDiskVolumeSource){PDName:(string)e2e-soak-486ab5a2-2487-11e5-b9d9-42010af01555 FSType:(string)ext4 Partition:(int)0 ReadOnly:(bool)true} AWSElasticBlockStore:(*api.AWSElasticBlockStoreVolumeSource)<nil> GitRepo:(*api.GitRepoVolumeSource)<nil> Secret:(*api.SecretVolumeSource)<nil> NFS:(*api.NFSVolumeSource)<nil> ISCSI:(*api.ISCSIVolumeSource)<nil> Glusterfs:(*api.GlusterfsVolumeSource)<nil> PersistentVolumeClaim:(*api.PersistentVolumeClaimVolumeSource)<nil> RBD:(*api.RBDVolumeSource)<nil>}} {Name:(string)default-token-812kd VolumeSource:(api.VolumeSource){HostPath:(*api.HostPathVolumeSource)<nil> EmptyDir:(*api.EmptyDirVolumeSource)<nil> GCEPersistentDisk:(*api.GCEPersistentDiskVolumeSource)<nil> AWSElasticBlockStore:(*api.AWSElasticBlockStoreVolumeSource)<nil> GitRepo:(*api.GitRepoVolumeSource)<nil> Secret:(*api.SecretVolumeSource){SecretName:(string)default-token-812kd} NFS:(*api.NFSVolumeSource)<nil> ISCSI:(*api.ISCSIVolumeSource)<nil> Glusterfs:(*api.GlusterfsVolumeSource)<nil> PersistentVolumeClaim:(*api.PersistentVolumeClaimVolumeSource)<nil> RBD:(*api.RBDVolumeSource)<nil>}}] Containers:([]api.Container)[{Name:(string)testpd Image:(string)gcr.io/google_containers/pause Command:([]string)<nil> Args:([]string)<nil> WorkingDir:(string) Ports:([]api.ContainerPort)<nil> Env:([]api.EnvVar)<nil> Resources:(api.ResourceRequirements){Limits:(api.ResourceList)map[cpu:100m] Requests:(api.ResourceList)<nil>} VolumeMounts:([]api.VolumeMount)[{Name:(string)testpd ReadOnly:(bool)false MountPath:(string)/testpd} {Name:(string)default-token-812kd ReadOnly:(bool)true MountPath:(string)/var/run/secrets/kubernetes.io/serviceaccount}] LivenessProbe:(*api.Probe)<nil> ReadinessProbe:(*api.Probe)<nil> Lifecycle:(*api.Lifecycle)<nil> TerminationMessagePath:(string)/dev/termination-log ImagePullPolicy:(api.PullPolicy)IfNotPresent SecurityContext:(*api.SecurityContext)<nil>}] RestartPolicy:(api.RestartPolicy)Always TerminationGracePeriodSeconds:(*int64)<nil> ActiveDeadlineSeconds:(*int64)<nil> DNSPolicy:(api.DNSPolicy)ClusterFirst NodeSelector:(map[string]string)<nil> ServiceAccountName:(string)default NodeName:(string)e2e-soak-minion-urqn HostNetwork:(bool)false ImagePullSecrets:([]api.LocalObjectReference)<nil>} Status:(api.PodStatus){Phase:(api.PodPhase)Failed Conditions:([]api.PodCondition)<nil> Message:(string)Pod cannot be started due to exceeded capacity Reason:(string)CapacityExceeded HostIP:(string) PodIP:(string) StartTime:(*util.Time)2015-07-07 02:05:56 -0700 PDT ContainerStatuses:([]api.ContainerStatus)<nil>}}",
    }
    Giving up; pod went into failed status: 
    (*api.Pod){TypeMeta:(api.TypeMeta){Kind:(string) APIVersion:(string)} ObjectMeta:(api.ObjectMeta){Name:(string)pd-test-4e63f3d7-2487-11e5-b9d9-42010af01555 GenerateName:(string) Namespace:(string)default SelfLink:(string)/api/v1/namespaces/default/pods/pd-test-4e63f3d7-2487-11e5-b9d9-42010af01555 UID:(types.UID)60556283-2487-11e5-a4e3-42010af01857 ResourceVersion:(string)15075 Generation:(int64)0 CreationTimestamp:(util.Time)2015-07-07 02:05:56 -0700 PDT DeletionTimestamp:(*util.Time)<nil> Labels:(map[string]string)<nil> Annotations:(map[string]string)<nil>} Spec:(api.PodSpec){Volumes:([]api.Volume)[{Name:(string)testpd VolumeSource:(api.VolumeSource){HostPath:(*api.HostPathVolumeSource)<nil> EmptyDir:(*api.EmptyDirVolumeSource)<nil> GCEPersistentDisk:(*api.GCEPersistentDiskVolumeSource){PDName:(string)e2e-soak-486ab5a2-2487-11e5-b9d9-42010af01555 FSType:(string)ext4 Partition:(int)0 ReadOnly:(bool)true} AWSElasticBlockStore:(*api.AWSElasticBlockStoreVolumeSource)<nil> GitRepo:(*api.GitRepoVolumeSource)<nil> Secret:(*api.SecretVolumeSource)<nil> NFS:(*api.NFSVolumeSource)<nil> ISCSI:(*api.ISCSIVolumeSource)<nil> Glusterfs:(*api.GlusterfsVolumeSource)<nil> PersistentVolumeClaim:(*api.PersistentVolumeClaimVolumeSource)<nil> RBD:(*api.RBDVolumeSource)<nil>}} {Name:(string)default-token-812kd VolumeSource:(api.VolumeSource){HostPath:(*api.HostPathVolumeSource)<nil> EmptyDir:(*api.EmptyDirVolumeSource)<nil> GCEPersistentDisk:(*api.GCEPersistentDiskVolumeSource)<nil> AWSElasticBlockStore:(*api.AWSElasticBlockStoreVolumeSource)<nil> GitRepo:(*api.GitRepoVolumeSource)<nil> Secret:(*api.SecretVolumeSource){SecretName:(string)default-token-812kd} NFS:(*api.NFSVolumeSource)<nil> ISCSI:(*api.ISCSIVolumeSource)<nil> Glusterfs:(*api.GlusterfsVolumeSource)<nil> PersistentVolumeClaim:(*api.PersistentVolumeClaimVolumeSource)<nil> RBD:(*api.RBDVolumeSource)<nil>}}] Containers:([]api.Container)[{Name:(string)testpd Image:(string)gcr.io/google_containers/pause Command:([]string)<nil> Args:([]string)<nil> WorkingDir:(string) Ports:([]api.ContainerPort)<nil> Env:([]api.EnvVar)<nil> Resources:(api.ResourceRequirements){Limits:(api.ResourceList)map[cpu:100m] Requests:(api.ResourceList)<nil>} VolumeMounts:([]api.VolumeMount)[{Name:(string)testpd ReadOnly:(bool)false MountPath:(string)/testpd} {Name:(string)default-token-812kd ReadOnly:(bool)true MountPath:(string)/var/run/secrets/kubernetes.io/serviceaccount}] LivenessProbe:(*api.Probe)<nil> ReadinessProbe:(*api.Probe)<nil> Lifecycle:(*api.Lifecycle)<nil> TerminationMessagePath:(string)/dev/termination-log ImagePullPolicy:(api.PullPolicy)IfNotPresent SecurityContext:(*api.SecurityContext)<nil>}] RestartPolicy:(api.RestartPolicy)Always TerminationGracePeriodSeconds:(*int64)<nil> ActiveDeadlineSeconds:(*int64)<nil> DNSPolicy:(api.DNSPolicy)ClusterFirst NodeSelector:(map[string]string)<nil> ServiceAccountName:(string)default NodeName:(string)e2e-soak-minion-urqn HostNetwork:(bool)false ImagePullSecrets:([]api.LocalObjectReference)<nil>} Status:(api.PodStatus){Phase:(api.PodPhase)Failed Conditions:([]api.PodCondition)<nil> Message:(string)Pod cannot be started due to exceeded capacity Reason:(string)CapacityExceeded HostIP:(string) PodIP:(string) StartTime:(*util.Time)2015-07-07 02:05:56 -0700 PDT ContainerStatuses:([]api.ContainerStatus)<nil>}}
not to have occurred

```

**Comment by bprashanth** [comment]
It's failing a fitness predicate on the node. 2 possibilities come to mind:
1. Race with static pod creation, where the scheduler assigned the pod as the mirror pod gets created, and the kubelet rejects it based on the new static pod.
2. Wrongfully reported node capacity, where the node reports 0, 0 limits based on which the scheduler happily assigns pods (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2007).

Otherwise, the same fitness predicate should fail on the scheduler side. Looks like we don't collect logs for the soak test, so I can't confirm either.

**Comment by davidopp** [comment]
Yeah, my changes to the priority function should not have anything to do with this.

**Comment by davidopp** [comment]
I didn't look at the test yet, but is there more than one node? If so, why didn't the pod reschedule onto another node once the resources for the mirror pod showed up?

**Comment by bprashanth** [comment]
It should, I think the test fails because one of the pods end up in Failed. In reality the rc would recreate it, and the scheduler would reassign to free host.

**Comment by davidopp** [comment]
So why didn't that happen?

**Comment by bprashanth** [comment]
The test failed based on the pod status https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/util.go#L471

**Comment by saad-ali** [comment]
@brendanburns noticed this issue 28 days ago: https://github.com/GoogleCloudPlatform/kubernetes/issues/9312#issuecomment-110784030

Meaning it's unlikely to be introduced recently.

**Comment by davidopp** [comment]
It would be good to see if this is happening soon after the cluster is created, or in steady-state. If it's the former, then it seems very likely the same as #10397. If it's the latter, then the cause might be different.

**Comment by saad-ali** [comment]
Doesn't appear to be happen *immediately* after cluster is created.
Cluster was created on Jul 7, 2015 at 2:44 PM
Bunch of E2E tests are run and pass.
This failure happens on Jul 7, 2015 at 3:49:07 PM

**Comment by yujuhong** [comment]
> It should, I think the test fails because one of the pods end up in Failed (unconfirmed, but judging from the message above). In reality the rc would recreate it, and the scheduler would reassign to free host.

The test does not use RC. It creates pods and writes the spec.NodeName directly. It bypasses scheduler completely.

The problem becomes why the node has so many pods running that it couldn't admit one more pod.

**Comment by yujuhong** [comment]
> The problem becomes why the node has so many pods running that it couldn't admit one more pod.

There a few possibilities that I can think of:
 1. One node was unready at one point, causing most pods to be shifted to the other node.
 2. Pods were not cleaned up in previous tests. This should not happen for tests using a non-default namespace (and with no resources specified in the pod spec).

Either way, logging the pods on the node would help.

**Comment by bprashanth** [comment]
Also suggest moving the test out of default namespace

**Comment by wojtek-t** [comment]
>
Also suggest moving the test out of default namespace

+1 - I will work on that.

**Comment by saad-ali** [comment]
I'm afraid that we might just be making an underlying issue by moving the test out of the default namespace. I'd still like to understand what's causing the issue.

I ran this test on my on GCE cluster back to back for several hours and never ran into the issue.

The Jenkins soak test, however, serially runs the entire E2E suite when it runs in to this issue. So maybe, as @yujuhong suggested, one of the other tests is not doing clean up properly. Or worse, there really is really a bug in the code.

To further debug, I will try to get a repro of this on my own GCE cluster, with additional logging by continuously running the full E2E suite.

**Comment by bprashanth** [comment]
There's a chance the bug is just 1. You can detect it by dumping node events eg: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/util.go#L1160 
You can simulate it by stopping the kubelet in a loop till the node controller deletes all pods on the node. The most likely reason for this happening is a leak somewhere and restarting the kubelet every 5m probably doesn't help because that's also how long the node controller waits before deleting pods.

**Comment by wojtek-t** [comment]
I agree this is probably 1 - we observed Kubelets restarting at random times in e2e tests - see #10792 for more details - e.g. https://github.com/GoogleCloudPlatform/kubernetes/issues/10792#issuecomment-119752028

**Comment by thockin** [comment]
Status update?

**Comment by saad-ali** [comment]
I ran the full E2E suite repeatedly overnight on my own GCE with additional logging around capacity exceeded logic, I was unable to repro the issue. I am dropping the priority of this issue, because if it is not blocking: even if it happens on a production system, the scheduler will reschedule the pod to a different node (whereas the PD test fails the test immediately).

**Comment by davidopp** [comment]
Unless they're not running under an RC, I guess? But that's probably rare.

**Comment by saad-ali** [comment]
Ok, I was able to capture some logs of this happening:

`plugin/pkg/scheduler/algorithm/predicates/predicates.go`
```
I0710 23:14:23.404760    3057 predicates.go:129] cannot fit pod ("default"/"66c106ea-2759-11e5-959f-42010af01f6c"). 
  fitsCPU=false
    totalMilliCPU=1000 milliCPURequested=1000 podRequest.milliCPU=100.
  fitsMemory=true
    totalMemory=3892027392 memoryRequested=524288000 podRequest.memory=0.
```

**Comment by bprashanth** [comment]
Yes, so that means the pod didn't fit on the node. did you happen to get a dump of pods currently on the node/do you have the cluster running?

**Comment by saad-ali** [comment]
Yep:
```
Logging pods the kubelet thinks is on node soak-gce-weekly-minion-otq4
INFO: elasticsearch-logging-v1-1em6o started at <nil> (0 container statuses recorded)
INFO: fluentd-elasticsearch-soak-gce-weekly-minion-otq4 started at <nil> (0 container statuses recorded)
INFO: elasticsearch-logging-v1-7h98v started at <nil> (0 container statuses recorded)
INFO: kube-ui-v1-q86il started at <nil> (0 container statuses recorded)
INFO: kibana-logging-v1-avym2 started at <nil> (0 container statuses recorded)
INFO: kube-dns-v6-xd167 started at <nil> (0 container statuses recorded)
INFO: monitoring-influx-grafana-v1-obzsj started at <nil> (0 container statuses recorded)
```

**Comment by saad-ali** [comment]
There were 7 pods running on the node:
```
E0710 23:14:23.404830    3057 kubelet.go:1585]  fitting: [7 pods] and non-fitting: [1 pod].
Machine capcity: map[cpu:{1.000 DecimalSI} memory:{3892027392 BinarySI}].
```
And the 7 pods had 10 containers total:
```
saadali@soak-gce-weekly-master:~$ kubectl get pods -o wide --all-namespaces | grep otq4
kube-system                       elasticsearch-logging-v1-1em6o                        1/1       Running   0          1h        soak-gce-weekly-minion-otq4
kube-system                       elasticsearch-logging-v1-7h98v                        1/1       Running   0          53m       soak-gce-weekly-minion-otq4
kube-system                       fluentd-elasticsearch-soak-gce-weekly-minion-otq4     1/1       Running   0          1h        soak-gce-weekly-minion-otq4
kube-system                       kibana-logging-v1-avym2                               1/1       Running   0          1h        soak-gce-weekly-minion-otq4
kube-system                       kube-dns-v6-xd167                                     3/3       Running   0          53m       soak-gce-weekly-minion-otq4
kube-system                       kube-ui-v1-q86il                                      1/1       Running   0          1h        soak-gce-weekly-minion-otq4
kube-system                       monitoring-influx-grafana-v1-obzsj                    2/2       Running   0          1h        soak-gce-weekly-minion-otq4
```
Each asking for `100m` so the `milliCPURequested=1000` makes sense.

The question then is why were there so many pods scheduled on one node.

@bprashanth suggested checking for evictions, and sure enough, the other node evicted pods:
```
saadali@soak-gce-weekly-master:~$ cat /var/log/kube-controller-manager.log | grep -i "podevict"
I0710 23:09:57.869452       9 podevictor.go:110] PodEvictor is evicting Pods on Node: soak-gce-weekly-minion-0u8y
I0710 23:51:40.474339       9 podevictor.go:110] PodEvictor is evicting Pods on Node: soak-gce-weekly-minion-0u8y
```

Now the question is what caused the other node to evict?

**Comment by davidopp** [comment]
If the other node evicted and the pods didn't end up back on that node, I think almost certainly they got evicted because the node became not-ready. Spreading would have put them back on the node that evicted if it was still healthy.

**Comment by saad-ali** [comment]
Ok, so this all makes sense now.

There is a `Network when a minion node becomes unreachable` test that runs before this PD test. It forces one node to go offline, which causes all the system pods to get rescheduled to the other node. And then the PD test tries to schedule its pod on the over-committed node and freaks out when it is unable to.

So changing the default namespace for the PD test pods should fix this. But, a better fix might be to somehow trigger a rebalance when starting the PD test. @bprashanth suggested maybe deleting and  existing RCs.

tldr: expected behavior/test issue

```
[0mNodes[0m [90mNetwork[0m [0mwhen a minion node becomes unreachable[0m 
  [1m[replication controller] recreates pods scheduled on the unreachable minion node AND allows scheduling of pods on a minion after it rejoins the cluster[0m
  [37m/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/resize_nodes.go:557[0m
[BeforeEach] Nodes
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/resize_nodes.go:412
>>> testContext.KubeConfig: /var/lib/jenkins/jobs/kubernetes-soak-weekly-deploy/workspace/.kube/config
INFO: Waiting up to 2m0s for service account default to be provisioned in ns e2e-tests-resize-nodes-ou7zq
INFO: Get service account default in ns e2e-tests-resize-nodes-ou7zq failed, ignoring for 5s: serviceaccounts "default" not found
INFO: Service account default in ns e2e-tests-resize-nodes-ou7zq with secrets found. (5.015614174s)
[BeforeEach] when a minion node becomes unreachable
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/resize_nodes.go:506
[It] [replication controller] recreates pods scheduled on the unreachable minion node AND allows scheduling of pods on a minion after it rejoins the cluster
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/resize_nodes.go:557
[1mSTEP[0m: creating replication controller my-hostname-net
INFO: Pod name my-hostname-net: Found 0 pods out of 2
INFO: Pod name my-hostname-net: Found 2 pods out of 2
[1mSTEP[0m: ensuring each pod is running
INFO: Waiting up to 5m0s for pod my-hostname-net-m4bja                                   status to be running
INFO: Found pod 'my-hostname-net-m4bja' on node 'soak-gce-weekly-minion-0u8y'
INFO: Waiting up to 5m0s for pod my-hostname-net-pelkk                                   status to be running
INFO: Found pod 'my-hostname-net-pelkk' on node 'soak-gce-weekly-minion-otq4'
[1mSTEP[0m: trying to dial each unique pod
INFO: Controller my-hostname-net: Got expected result from replica 1 [my-hostname-net-m4bja]: "my-hostname-net-m4bja", 1 of 2 required successes so far
INFO: Controller my-hostname-net: Got expected result from replica 2 [my-hostname-net-pelkk]: "my-hostname-net-pelkk", 2 of 2 required successes so far
[1mSTEP[0m: choose a node with at least one pod - we will block some network traffic on this node
[1mSTEP[0m: block network traffic from node soak-gce-weekly-minion-0u8y
INFO: Getting external IP address for soak-gce-weekly-minion-0u8y
[1mSTEP[0m: block network traffic from node soak-gce-weekly-minion-0u8y to the master
INFO: Using SSH key: /var/lib/jenkins/jobs/kubernetes-soak-weekly-deploy/workspace/.ssh/google_compute_engine
INFO: Waiting for node soak-gce-weekly-minion-0u8y to be not ready
INFO: Waiting up to 2m0s for node soak-gce-weekly-minion-0u8y readiness to be false
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: True
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: True
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: True
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: True
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: True
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: True
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: True
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Successfully found node soak-gce-weekly-minion-0u8y readiness to be false
INFO: Waiting for pod my-hostname-net-m4bja to be removed
INFO: Waiting for pod my-hostname-net-m4bja to disappear
INFO: Pod my-hostname-net-m4bja still exists
INFO: Waiting for pod my-hostname-net-m4bja to disappear
INFO: Pod my-hostname-net-m4bja still exists
INFO: Waiting for pod my-hostname-net-m4bja to disappear
INFO: Pod my-hostname-net-m4bja still exists
INFO: Waiting for pod my-hostname-net-m4bja to disappear
INFO: Pod my-hostname-net-m4bja still exists
INFO: Waiting for pod my-hostname-net-m4bja to disappear
INFO: Pod my-hostname-net-m4bja still exists
INFO: Waiting for pod my-hostname-net-m4bja to disappear
INFO: Pod my-hostname-net-m4bja still exists
INFO: Waiting for pod my-hostname-net-m4bja to disappear
INFO: Pod my-hostname-net-m4bja still exists
INFO: Waiting for pod my-hostname-net-m4bja to disappear
INFO: Pod my-hostname-net-m4bja still exists
INFO: Waiting for pod my-hostname-net-m4bja to disappear
INFO: Pod my-hostname-net-m4bja still exists
INFO: Waiting for pod my-hostname-net-m4bja to disappear
INFO: Pod my-hostname-net-m4bja still exists
INFO: Waiting for pod my-hostname-net-m4bja to disappear
INFO: Pod my-hostname-net-m4bja no longer exists
[1mSTEP[0m: verifying whether the pod from the unreachable node is recreated
INFO: Pod name my-hostname-net: Found 2 pods out of 2
[1mSTEP[0m: ensuring each pod is running
INFO: Waiting up to 5m0s for pod my-hostname-net-aaqma                                   status to be running
INFO: Waiting for pod my-hostname-net-aaqma                                   in namespace 'e2e-tests-resize-nodes-ou7zq' status to be 'running'(found phase: "Pending", readiness: false) (3.928301ms elapsed)
INFO: Found pod 'my-hostname-net-aaqma' on node 'soak-gce-weekly-minion-otq4'
INFO: Waiting up to 5m0s for pod my-hostname-net-pelkk                                   status to be running
INFO: Found pod 'my-hostname-net-pelkk' on node 'soak-gce-weekly-minion-otq4'
[1mSTEP[0m: trying to dial each unique pod
INFO: Controller my-hostname-net: Got expected result from replica 1 [my-hostname-net-aaqma]: "my-hostname-net-aaqma", 1 of 2 required successes so far
INFO: Controller my-hostname-net: Got expected result from replica 2 [my-hostname-net-pelkk]: "my-hostname-net-pelkk", 2 of 2 required successes so far
[1mSTEP[0m: Unblock network traffic from node soak-gce-weekly-minion-0u8y to the master
INFO: Using SSH key: /var/lib/jenkins/jobs/kubernetes-soak-weekly-deploy/workspace/.ssh/google_compute_engine
INFO: Waiting for node soak-gce-weekly-minion-0u8y to be ready
INFO: Waiting up to 2m0s for node soak-gce-weekly-minion-0u8y readiness to be true
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: Unknown
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: True
INFO: Successfully found node soak-gce-weekly-minion-0u8y readiness to be true
[1mSTEP[0m: verify whether new pods can be created on the re-attached node
INFO: Created pod additionalpod on node soak-gce-weekly-minion-0u8y
INFO: Pod name additionalpod: Found 1 pods out of 1
[1mSTEP[0m: ensuring each pod is running
INFO: Waiting up to 5m0s for pod additionalpod                                           status to be running
INFO: Waiting for pod additionalpod                                           in namespace 'e2e-tests-resize-nodes-ou7zq' status to be 'running'(found phase: "Pending", readiness: false) (20.470499ms elapsed)
INFO: Found pod 'additionalpod' on node 'soak-gce-weekly-minion-0u8y'
[1mSTEP[0m: trying to dial each unique pod
INFO: Controller additionalpod: Got expected result from replica 1 [additionalpod]: "additionalpod", 1 of 1 required successes so far
[AfterEach] Nodes
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/resize_nodes.go:423
[1mSTEP[0m: checking whether all nodes are healthy
INFO: Node soak-gce-weekly-minion-0u8y condition 1/1: type: Ready, status: True
INFO: Node soak-gce-weekly-minion-otq4 condition 1/1: type: Ready, status: True
[1mSTEP[0m: destroying namespace for this suite e2e-tests-resize-nodes-ou7zq

[32m• [SLOW TEST:376.324 seconds][0m
Nodes
[90m/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/resize_nodes.go:560[0m
  Network
  [90m/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/resize_nodes.go:559[0m
    when a minion node becomes unreachable
    [90m/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/resize_nodes.go:558[0m
      [replication controller] recreates pods scheduled on the unreachable minion node AND allows scheduling of pods on a minion after it rejoins the cluster
      [90m/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/resize_nodes.go:557[0m

```

**Comment by bprashanth** [comment]
> suggested maybe deleting and existing RCs.

Deleting existing pods in kube-system ns. Deleting the rcs will cause more problems. Unsure if this is required, but unbalanced cluster addons might lead to issues for other tests.

**Comment by quinton-hoole** [comment]
Awesome @saad-ali! Well sleuthed.
I don't think (or understand why) moving that test to it's own namespace would make any difference?
I think that we should consider the "Network when a minion node becomes unreachable" test to be potentially cluster-destructive on small two-node clusters like out test clusters (it would be a relative non-event on larger clusters, where the evicted pods would be spread around much better).  IMO, that test should either run in it's own test cluster, or together with the other potentially destructive "Reboot" tests (with which it might also interfere, or course).

**Comment by quinton-hoole** [comment]
Or we could routinely build larger (3-4 non-master node) clusters for testing purposes.  This is not the first time we've come across a case where a 2-node cluster is simply not representative of reality, and produces false-positives (er.. negatives.. whatever).

**Comment by davidopp** [comment]
Thanks for debugging, @saad-ali !

I agree with @quinton-hoole regarding fixes.

**Comment by thockin** [comment]
I agree that 2 nodes is not realistic for the full gamut of tests we want to run.  @quinton-hoole is away and I don't know that much Jenkins, but we should probably bump to to 3 right away.  Disagreements? 

@ixdy - do you know how, assuming we do this?

**Comment by wojtek-t** [comment]
@thockin - changing the cluster size is a trivial change. All we need to do is to set "NUM_NIONS to 3 (or 4) instead of 2:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/hack/jenkins/e2e.sh#L54

**Comment by thockin** [comment]
Someone send a PR?

**Comment by eikaas** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  tapiohush@gmail.com
    Domain eikaas.no has exceeded the max emails per hour (33/30 (110%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from o6.sgmail.github.com ([192.254.113.101]:52492)
	by cpanel30.proisp.no with esmtps (TLSv1.2:DHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.85)
	(envelope-from <bounces+848413-7097-robin=eikaas.no@sgmail.github.com>)
	id 1ZEgoE-00429A-GY
	for robin@eikaas.no; Mon, 13 Jul 2015 18:42:28 +0200
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed; d=github.com; 
	h=from:reply-to:to:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=K7C1USUbakzwpZXvtWouAcgOA6I=; b=T8NkYm2/y31JIZMR
	VZe9kX6ZtvrATlDQlpOFOnmoBquXy/C9nWzsyzPokgMap/zssxkL1r/zXlkzlY9J
	dQIEd5cV3cxB+rvXlQfH9hRgqgcGdOHtRZU0savUEShu8UShsF7As7WelmHYftGP
	bcYsNaBLDfsuStglWAoMDZUGkWA=
Received: by filter0603p1mdw1.sendgrid.net with SMTP id filter0603p1mdw1.28133.55A3EA6DB
        2015-07-13 16:42:22.046785187 +0000 UTC
Received: from github-smtp2a-ext-cp1-prd.iad.github.net (github-smtp2a-ext-cp1-prd.iad.github.net [192.30.253.16])
	by ismtpd-023 (SG) with ESMTP id 14e884bbdca.76ef.27fc13
	for <robin@eikaas.no>; Mon, 13 Jul 2015 16:42:22 +0000 (UTC)
Date: Mon, 13 Jul 2015 09:42:21 -0700

**Comment by brendandburns** [comment]
https://github.com/GoogleCloudPlatform/kubernetes/pull/11165 sent.

**Comment by saad-ali** [comment]
Pushing a new build to soak to verify #11165

**Comment by saad-ali** [comment]
Did a clean deployment to Jenkins GCE Soak at 4PM yesterday. All runs since have been green. Closing this issue.
